Add handle-release-branches-workflow

kylegach opened the pull request:
## Why

To trigger branch-based versioned docs builds, we need to do some branch management between this repo and [`frontpage`](https://github.com/storybookjs/frontpage). By automating the creation of the `release-x-x` branches (both here and in `frontpage`), we can develop as usual, without having to manually ensure that the `release-x-x` branches are kept in-sync with `main` & `next` (here) and `master` (frontpage).

## What I did

### Scenarios

> For each of the scenarios below, assume:
> - "latest" (`master` branch) version is 6.3
> - "next" (`next` branch) version is 6.4

#### On push to `main`

- Use webhook to kick off production frontpage deploy

![main](https://user-images.githubusercontent.com/486540/140180927-27eb3f78-7198-4505-b9d3-84e9be0c80fa.png)

#### On push to `next`

- Creates & force-pushes `release-6-4` branch
- Sends dispatch event to frontpage repo to create `release-6-4` branch

![next](https://user-images.githubusercontent.com/486540/140180933-c94468e8-57a6-456f-bc72-87705e8dc34a.png)

#### On push to `release-x-x`

- If pushing to `release-6-4`
    - Warns that changes will be lost on next push to `next`
    - ![release-6-3](https://user-images.githubusercontent.com/486540/140180936-0003cb33-9c51-4bb6-9ded-87979b3c113f.png)
    - ![release-6-3-error](https://user-images.githubusercontent.com/486540/140180939-8779eefd-28eb-4f38-99c8-ad214e9806ed.png)
- Else
    - Sends dispatch event to frontpage repo to create `release-x-x` branch
    - ![release-6-0](https://user-images.githubusercontent.com/486540/140180935-5edd1350-bac1-4c7f-94df-e3100dfe65e1.png)

## Questions

- Should this PR be against `main`?
- I _think_ this workflow would obviate the need for the [`build-frontpage` script](https://github.com/storybookjs/storybook/blob/next/scripts/build-frontpage.js). Is that correct?
    - If not, that script or this workflow will likely need modified to live alongside one another
- Have I done enough to safeguard this workflow?
- Is there a better way to alert of potentially lost changes than a failure w/ logged warning?
    - PR comment, perhaps?
- Sending the dispatch event to `frontpage` will require an access token. I'm unsure what that should be. Do we have a bot account for such things?

kylegach commented on .github/workflows/handle-release-branches.yml:
```diff
@@ -0,0 +1,126 @@
+    steps:
+      - if: ${{ needs.branch-checks.outputs.is-actionable-branch == 'true' && needs.latest-or-next-release-branch-check.outputs.check == 'false' }}
+        run: |
+          curl -X POST https://api.github.com/repos/storybookjs/frontpage/dispatches \
```
This technique is from https://blog.marcnuri.com/triggering-github-actions-across-different-repositories.

The `frontpage` workflow would look like:

```yml
name: Respond to request-create-frontpage-branch

on: [repository_dispatch]

jobs:
  build:
    runs-on: ubuntu-latest
    if: github.event.action == 'request-create-frontpage-branch'
    steps:
      - uses: actions/checkout@v2
      - run: |
          echo "Creating `${{ github.event.client_payload.branch }}` branch..."
          git branch temp
          git push -f origin temp:${{ github.event.client_payload.branch }}
          curl -X POST https://api.github.com/repos/kylegach/actions-test-a/dispatches \
          -H 'Accept: application/vnd.github.v3+json' \
          -u ${{ secrets.ACCESS_TOKEN }} \
          --data '{"event_type": "created-frontpage-branch", "client_payload": { "branch": "${{ needs.branch-checks.outputs.branch }}" }}'
```

Pushing the `release-x-x` branch would then kick off a Netlify build for that branch, which would pull in the appropriate version's content from this repo's `release-x-x` branch.

The `curl` above would allow for _this_ repo to receive a confirmation message, if desired. Otherwise, it can be removed.

shilman commented on .github/workflows/handle-release-branches.yml:
```diff
@@ -0,0 +1,126 @@
+      - uses: actions/checkout@v2
+      - run: |
+          git branch temp
+          git push -f origin temp:${{ needs.get-next-release-branch.outputs.branch }}
```
Why `temp.xxx`?

shilman reviewed (commented):
> Should this PR be against main?

No, `next` is the default branch

> I think this workflow would obviate the need for the build-frontpage script. Is that correct?
> If not, that script or this workflow will likely need modified to live alongside one another

Yes, I believe so

> Have I done enough to safeguard this workflow? Is there a better way to alert of potentially lost changes than a failure w/ logged warning? Which steps could fail?

At any rate, since this is only copying the branch automatically, we never lose information correct? We only run the risk of some info not propagating properly

> Sending the dispatch event to frontpage will require an access token. I'm unsure what that should be. Do we have a bot account for such things?

Yes we do -- DM'ing

kylegach commented:
> Why temp.xxx?

Because the value of `${{ needs.get-next-release-branch.outputs.branch }}`, e.g. `release-6-4` is likely to already exist, and this was easier and more straightforward (imo) than finding a way around the error when creating a branch that already exists.

shilman commented:
@kylegach is the frontpage side set up to read `temp:xxx` branches? or am i misunderstanding

kylegach commented:
> @kylegach is the frontpage side set up to read `temp:xxx` branches? or am i misunderstanding

That's a syntax to push the local `temp` branch to (something like) `release-6-4` on the remote.

More info: https://stackoverflow.com/a/28302850

kylegach commented:
> since this is only copying the branch automatically, we never lose information correct? We only run the risk of some info not propagating properly

Realized I never answered this. 🤦‍♂️ To explain, I'll keep the same assumptions from the description: "latest" = `6.3` and "next" = `6.4`. This workflow has one potentially dangerous outcome: push to `next` creates `release-6-4`, which is then force-pushed over whatever `release-6-4` may be on remote. So I'm fairly sure the only data loss would be from someone pushing to `release-6-4` directly, which this workflow explicitly no-ops and warns about (via a failure and a log message).

kylegach commented on .github/workflows/handle-release-branches.yml:
```diff
@@ -0,0 +1,126 @@
+    outputs:
+      branch: ${{ env.latest-release-branch }}
+  
+  create-latest-release-branch:
```
Rather than creating a branch for this case, it would be far more straightforward^1 to just use the Netlify webhook directly (as is currently done in the `build-frontpage` script).

[1] The latest version of the docs will be the non-versioned production build of the site, so trying to build that from a branch other than `master` would be more complex than necessary.

kylegach commented on .circleci/config.yml:
```diff
@@ -353,17 +353,6 @@ jobs:
           command: |
             cd examples/cra-react15
             yarn storybook --smoke-test --quiet
-  frontpage:
```
Please ensure I removed this correctly, as I'm not all that familiar with CircleCI. (Same with TeamCity changes.) 🙏

kylegach commented on .github/workflows/handle-release-branches.yml:
```diff
@@ -0,0 +1,124 @@
+    steps:
+      - uses: actions/checkout@v2
+
+      - run: curl -X POST "https://api.netlify.com/build_hooks/$FRONTPAGE_HOOK"
```
Will likely need to add `FRONTPAGE_HOOK` env var to this repo.

kylegach commented:
I don't believe the failing CircleCi checks, which are both e2e tests, are related.

kylegach committed:
Add handle-release-branches-workflow

- Assuming "latest" is 6.3 & "next" is 6.4...
- On push to `main`
    - Use webhook to kick off production frontpage deploy
- On push to `next`
    - Creates & force-pushes `release-6-4` branch
    - Sends dispatch event to frontpage repo to create `release-6-4` branch
- On push to `release-x-x`
    - If pushing to `release-6-4`
        - Warns that changes will be lost on next push to `next`
    - Else
        - Sends dispatch event to frontpage repo to create `release-n-n` branch
- Remove `build-frontpage` script
    - Remove references from teamcity & circleci configs

shilman commented:
Amazing job @kylegach !!! 🚀
